core/drivers/bcm_sotp.c

```diff
@@ -0,0 +1,131 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2019 Broadcom.
+ */
+
+#include <assert.h>
+#include <drivers/bcm_sotp.h>
+#include <initcall.h>
+#include <io.h>
+#include <kernel/delay.h>
+#include <mm/core_memprot.h>
+#include <platform_config.h>
+#include <util.h>
+
+#define SOTP_PROG_CONTROL			0x0
+#define SOTP_PROG_CONTROL__OTP_CPU_MODE_EN	BIT(15)
+#define SOTP_PROG_CONTROL__OTP_DISABLE_ECC	BIT(9)
+#define SOTP_ADDR__OTP_ROW_ADDR_R		6
+
+#define SOTP_ADDR				0xc
+
+#define SOTP_CTRL_0				0x10
+#define SOTP_CTRL_0__START			1
+#define SOTP_READ				0
+
+#define SOTP_STAT_0				0x18
+#define SOTP_STATUS_0__FDONE			BIT(3)
+
+#define SOTP_STATUS_1				0x1c
+#define SOTP_STATUS_1__CMD_DONE			BIT(1)
+#define SOTP_STATUS_1__ECC_DET			BIT(17)
+
+#define SOTP_RDDATA_0				0x20
+#define SOTP_RDDATA_1				0x24
+#define SOTP_ADDR_MASK				0x3ff
+
+#define SOTP_ECC_ERR_DETECT			BIT64(63)
+
+#define SOTP_TIMEOUT_US				300
+
+static vaddr_t bcm_sotp_base;
+
+static TEE_Result otp_status_done_wait(vaddr_t addr, uint32_t bit)
+{
+	uint64_t timeout = timeout_init_us(SOTP_TIMEOUT_US);
+
+	while (!(io_read32(addr) & bit))
+		if (timeout_elapsed(timeout))
+			return TEE_ERROR_BUSY;
+	return TEE_SUCCESS;
+}
+
+TEE_Result bcm_iproc_sotp_mem_read(uint32_t row_addr, uint32_t sotp_add_ecc,
+				uint64_t *rdata)
+{
+	uint64_t read_data = 0;
+	uint32_t reg_val = 0;
+	int ret = TEE_SUCCESS;
```

Comment:
Please use `TEE_Result` instead of `int`.

Reply:
done